pkg/diffs/diffs_test.go

```diff
@@ -0,0 +1,132 @@
+package diffs
+
+import (
+	"testing"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	prowconfig "k8s.io/test-infra/prow/config"
+)
+
+const orgRepo = "org/repo"
+
+func TestGetChangedPresubmits(t *testing.T) {
+	basePreSubmit := makeBasePresubmit()
+	testCases := []struct {
+		name                  string
+		masterConfigGenerator func() (config *prowconfig.Config)
+		prConfigGenerator     func() (config *prowconfig.Config)
+		expected              map[string][]prowconfig.Presubmit
+	}{
+		{
+			name: "no differences mean nothing is identified as a diff",
+			masterConfigGenerator: func() *prowconfig.Config {
+				return makeConfig(basePreSubmit)
+			},
+			prConfigGenerator: func() *prowconfig.Config {
+				return makeConfig(basePreSubmit)
+			},
+			expected: map[string][]prowconfig.Presubmit{},
+		},
+		{
+			name: "different agent is identified as a diff (from jenkins to kubernetes)",
+			masterConfigGenerator: func() *prowconfig.Config {
+				p := *basePreSubmit
+				p.Agent = "jenkins"
+				return makeConfig(&p)
+			},
+			prConfigGenerator: func() *prowconfig.Config {
+				return makeConfig(basePreSubmit)
+			},
+			expected: map[string][]prowconfig.Presubmit{
+				orgRepo: {*basePreSubmit},
+			},
+		},
+		{
+			name: "different spec is identified as a diff - single change",
+			masterConfigGenerator: func() *prowconfig.Config {
+				return makeConfig(basePreSubmit)
+			},
+			prConfigGenerator: func() *prowconfig.Config {
+				p := makeBasePresubmit()
+				p.Spec.Containers[0].Command = []string{"test-command"}
+				return makeConfig(p)
+			},
+			expected: map[string][]prowconfig.Presubmit{
+				orgRepo: {
+					func() prowconfig.Presubmit {
+						p := makeBasePresubmit()
+						p.Spec.Containers[0].Command = []string{"test-command"}
+						return *p
+					}()}},
+		},
+		{
+			name: "different spec is identified as a diff - massive changes",
+			masterConfigGenerator: func() *prowconfig.Config {
+				return makeConfig(basePreSubmit)
+			},
+			prConfigGenerator: func() *prowconfig.Config {
+				p := makeBasePresubmit()
+				p.Spec.Containers[0].Command = []string{"test-command"}
+				p.Spec.Containers[0].Args = []string{"testarg", "testarg", "testarg"}
+				p.Spec.Volumes = []v1.Volume{{
+					Name: "test-volume",
+					VolumeSource: v1.VolumeSource{
+						EmptyDir: &v1.EmptyDirVolumeSource{},
+					}},
+				}
+				return makeConfig(p)
+			},
+			expected: map[string][]prowconfig.Presubmit{
+				orgRepo: {
+					func() prowconfig.Presubmit {
+						p := makeBasePresubmit()
+						p.Spec.Containers[0].Command = []string{"test-command"}
+						p.Spec.Containers[0].Args = []string{"testarg", "testarg", "testarg"}
+						p.Spec.Volumes = []v1.Volume{{
+							Name: "test-volume",
+							VolumeSource: v1.VolumeSource{
+								EmptyDir: &v1.EmptyDirVolumeSource{},
+							}},
+						}
+						return *p
+					}()}},
+		},
+	}
+
+	for _, testCase := range testCases {
```

Comment:
Would like @petr-muller review on this before merge

Reply:
They'll run faster, too

Reply:
addressed